### PR TITLE
Move python tool configurations to pyproject.toml, and add the python 3.11 classifier.

### PR DESCRIPTION
**Pull request by UriyaHarpeness**
* Add missing `Programming Language :: Python :: 3.11` classifier for the python package.
* Move `pylint` config to the `pyproject.toml` and delete `.pylintrc`.
* Move `mypy` config to the `pyproject.toml` and delete `setup.cfg`.
* Improve `lint_python.py`:
  * Reuse and readability.
  * Stop using soon-to-be-deprecated `epylint`.
  * Use the flags from the `pyproject.toml` instead of passing them through the CLI.

### CODE REVIEW

**Comment by UriyaHarpeness** [comment]
Hey @trivialfis @hcho3,
While on this subject, I see some flags are being passed in https://github.com/dmlc/xgboost/blob/master/tests/ci_build/lint_python.py, can these flags be removed if they already exist in the `pyproject.toml`? With perhaps changing the CWD when running the linting script?
It's a shame to have several copies that could get misaligned.

**Comment by trivialfis** [comment]
For `mypy`, it should work out of the box. pylint however, is a bit more messy as we haven't upgraded the configuration for a while, the file you referenced is based on some really old and deprecated features from pylint. I can try to upgrade the configuraiton later.

**Comment by UriyaHarpeness** [comment]
@trivialfis please let me try to do it myself first, will do so very soon.

**Comment by trivialfis** [comment]
Thank you for the nice work! Let me know if there's anything I can help.

**Comment by UriyaHarpeness** [comment]
@trivialfis am I free to use the builtin library `pathlib`? Would also help simplify this linting script.

**Comment by trivialfis** [comment]
yup, whatever that comes with python.

**Comment by hcho3** [comment]
Thanks for working on updating the linting script. Let me know if there's anything I can help.

**Comment by UriyaHarpeness** [comment]
@hcho3 @trivialfis I think I'm done here, locally everything works.
The only thing I would've maybe done is split the actual linting script and it's configurations, currently it's mixed as seen below:
```python
...
        mypy_results = [
            run_mypy(path)
            for path in [
                # core
                "python-package/",
                # demo
                "demo/json-model/json_parser.py",
                "demo/guide-python/external_memory.py",
                "demo/guide-python/cat_in_the_dat.py",
...
```
I am very willing to do so, let me know if it's acceptable by you and I shall change it.

**Comment by UriyaHarpeness** [comment]
@hcho3 @trivialfis poke.

File: tests/ci_build/lint_python.py [hunk 1]
```diff
@@ -1,137 +1,133 @@
 import argparse
 import os
+import pathlib
 import subprocess
 import sys
+from collections import Counter
 from multiprocessing import Pool, cpu_count
-from typing import Dict, Tuple
+from typing import Dict, List, Tuple, cast
```
Comment (trivialfis) [inline tests/ci_build/lint_python.py:8]: Please remove the `cast` import, which is not used.

**Review by trivialfis — COMMENTED** [review]
Very nice! Now the repository is much more consistent and unified, along with the linting code being cleaned up.

One minor issue mentioned in the comment, otherwise looks good to me.

File: tests/ci_build/lint_python.py (hunk 1 above)
Comment (trivialfis) [reply tests/ci_build/lint_python.py:8]: For a small file like the config I think it's fine. The content is separated by sections.

Comment (UriyaHarpeness) [reply tests/ci_build/lint_python.py]: Removed the unused `cast`.
Also unified the linting paths under a class in the same file, helps a lot to see where the config vs code is.

**Comment by UriyaHarpeness** [comment]
I think the PR is done now, I'm pretty happy with it.
Now, if I'm the one that should merge it just give me your approvals and I shall do so, otherwise - please go ahead and merge if it's to your liking.
Thanks @trivialfis @hcho3.

**Review by trivialfis — APPROVED** [review]

**Comment by trivialfis** [comment]
thank you for your nice work on cleaning up the configuration and helper scripts!

**Comment by UriyaHarpeness** [comment]
My pleasure! glad I could contribute.
@trivialfis @hcho3

**Pull request merged**